File: api/src/test/java/org/pmiops/workbench/conceptset/ConceptSetServiceTest.java
package org.pmiops.workbench.conceptset;

import static com.google.common.truth.Truth.assertThat;

import java.util.Set;
import java.util.stream.Collectors;
import java.util.stream.Stream;
import org.junit.jupiter.api.Test;
import org.pmiops.workbench.FakeClockConfiguration;
import org.pmiops.workbench.cdr.ConceptBigQueryService;
import org.pmiops.workbench.cohortbuilder.CohortBuilderService;
import org.pmiops.workbench.cohortbuilder.mapper.CohortBuilderMapper;
import org.pmiops.workbench.conceptset.mapper.ConceptSetMapper;
import org.pmiops.workbench.conceptset.mapper.ConceptSetMapperImpl;
import org.pmiops.workbench.db.dao.WorkspaceDao;
import org.pmiops.workbench.db.model.DbConceptSet;
import org.pmiops.workbench.db.model.DbConceptSetConceptId;
import org.pmiops.workbench.db.model.DbUser;
import org.pmiops.workbench.db.model.DbWorkspace;
import org.pmiops.workbench.model.Domain;
import org.pmiops.workbench.utils.mappers.CommonMappers;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.orm.jpa.DataJpaTest;
import org.springframework.boot.test.context.TestConfiguration;
import org.springframework.boot.test.mock.mockito.MockBean;
import org.springframework.context.annotation.Import;
import org.springframework.test.annotation.DirtiesContext;

@DataJpaTest
@DirtiesContext(classMode = DirtiesContext.ClassMode.BEFORE_EACH_TEST_METHOD)
public class ConceptSetServiceTest {

  @Autowired WorkspaceDao workspaceDao;
  @Autowired ConceptSetService conceptSetService;
  @Autowired ConceptSetMapper conceptSetMapper;

  @TestConfiguration
  @Import({FakeClockConfiguration.class, ConceptSetService.class, ConceptSetMapperImpl.class})
  @MockBean({
    CommonMappers.class,
    CohortBuilderMapper.class,
    ConceptBigQueryService.class,
    CohortBuilderService.class,
    DbUser.class
  })
  static class Configuration {}

  @Test
  public void testCloneConceptSetWithNoCdrVersionChange() {
    DbWorkspace mockDbWorkspace = mockWorkspace();
    DbConceptSet fromConceptSet = mockConceptSet();
    DbConceptSet copiedConceptSet =
        conceptSetService.cloneConceptSetAndConceptIds(fromConceptSet, mockDbWorkspace);
    assertThat(copiedConceptSet).isNotNull();
    assertThat(copiedConceptSet.getConceptSetConceptIds().size()).isEqualTo(5);
    assertThat(copiedConceptSet.getWorkspaceId()).isEqualTo(mockDbWorkspace.getWorkspaceId());
    assertThat(copiedConceptSet).isNotNull();
  }

  private DbConceptSet mockConceptSet() {
    Set<DbConceptSetConceptId> dbConceptSetConceptIds =
        Stream.of(1L, 2L, 3L, 4L, 5L)
            .map(id -> DbConceptSetConceptId.builder().addConceptId(id).addStandard(true).build())
            .collect(Collectors.toSet());

    DbConceptSet conceptSet = new DbConceptSet();
    conceptSet.setConceptSetConceptIds(dbConceptSetConceptIds);
    conceptSet.setConceptSetId(1);
    conceptSet.setName("Mock Concept Set");
    conceptSet.setDomainEnum(Domain.CONDITION);
    return conceptSet;
  }

  private DbWorkspace mockWorkspace() {
    DbWorkspace workspace = new DbWorkspace();
    workspace.setName("Target DbWorkspace");
    workspace.setWorkspaceId(2);
    workspace = workspaceDao.save(workspace);
    return workspace;
  }
}

File: api/src/test/java/org/pmiops/workbench/conceptset/ConceptSetServiceTest.kt
package org.pmiops.workbench.conceptset

import org.junit.Assert.assertEquals
import org.junit.Assert.assertNotNull

import java.util.HashSet
import java.util.stream.Collectors
import java.util.stream.Stream
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.pmiops.workbench.db.dao.ConceptSetDao
import org.pmiops.workbench.db.dao.WorkspaceDao
import org.pmiops.workbench.db.model.ConceptSet
import org.pmiops.workbench.db.model.Workspace
import org.pmiops.workbench.model.Domain
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.autoconfigure.liquibase.LiquibaseAutoConfiguration
import org.springframework.boot.test.autoconfigure.jdbc.AutoConfigureTestDatabase
import org.springframework.boot.test.autoconfigure.orm.jpa.DataJpaTest
import org.springframework.context.annotation.Import
import org.springframework.test.annotation.DirtiesContext
import org.springframework.test.context.junit4.SpringRunner

@RunWith(SpringRunner::class)
@DataJpaTest
@Import(LiquibaseAutoConfiguration::class)
@AutoConfigureTestDatabase(replace = AutoConfigureTestDatabase.Replace.NONE)
@DirtiesContext(classMode = DirtiesContext.ClassMode.BEFORE_EACH_TEST_METHOD)
class ConceptSetServiceTest {

    @Autowired
    private val conceptSetDao: ConceptSetDao? = null

    @Autowired
    private val workspaceDao: WorkspaceDao? = null

    private var conceptSetService: ConceptSetService? = null
    private var workspace: Workspace? = null

    @Before
    fun setUp() {
        conceptSetService = ConceptSetService()
        conceptSetService!!.setConceptSetDao(conceptSetDao)
        workspace = mockWorkspace()
    }

    @Test
    fun testCloneConceptSetWithNoCdrVersionChange() {
        val fromConceptSet = mockConceptSet()
        val copiedConceptSet = conceptSetService!!.cloneConceptSetAndConceptIds(fromConceptSet, workspace, false)
        assertNotNull(copiedConceptSet)
        assertEquals(copiedConceptSet.conceptIds.size.toLong(), 5)
        assertEquals(copiedConceptSet.workspaceId, workspace!!.workspaceId)
    }

    private fun mockConceptSet(): ConceptSet {
        val conceptIdsSet = Stream.of(1, 2, 3, 4, 5).collect<HashSet<Int>, Any>(Collectors.toCollection(Supplier<HashSet<Int>> { HashSet() }))

        val conceptSet = ConceptSet()
        conceptSet.conceptIds = conceptIdsSet
        conceptSet.conceptSetId = 1
        conceptSet.name = "Mock Concept Set"
        conceptSet.domainEnum = Domain.CONDITION
        return conceptSet
    }

    private fun mockWorkspace(): Workspace {
        var workspace = Workspace()
        workspace.name = "Target Workspace"
        workspace.workspaceId = 2
        workspace = workspaceDao!!.save(workspace)
        return workspace
    }
}
